naming-conventions.md

```diff
@@ -0,0 +1,106 @@
+```haskell
+foldr :: Foldable t => (a -> b -> b) -> b -> t a -> b
+traverse :: (Applicative f, Traversable t) => (a -> f b) -> t a -> f (t b)
+```
+
+`m` is always always `Monad`
+```haskell
+when :: Monad m => Bool -> m () -> m ()
+```
+
+`s` is for state and `e` is for environment
+```haskell
+put :: MonadState s m => s -> m ()
+ask :: MonadReader e m => m e
+```
+
+These are conventions only, but picking something unusual for a common typeclass might surprise other readers.
+
+### Concrete Types
+
+Avoid using single-letter names or abbreviations for values with a concrete type, especially if that type is domain-specific.
+
+For example, don't do this:
+```haskell 
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity tId teacher) =
+  select $ from $ \(ts `InnerJoin` cs `InnerJoin` cms) -> do
+    on $ cs ^. CourseId ==. cms  ^. CourseMembershipCourseId
+    on $ cs ^. CourseTeacherId ==. ts ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just sId ->
+          ts ^. TeacherSchoolId ==. val (Just sId)
+        Nothing ->
+          ts ^. TeacherId ==. val tId
+    pure cms
+```
+
+Do this instead:
+```haskell
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity teacherId teacher) =
+  select $ from $ \(teachers `InnerJoin` courses `InnerJoin` memberships) -> do
+    on $ courses ^. CourseId ==. memberships  ^. CourseMembershipCourseId
+    on $ courses ^. CourseTeacherId ==. teachers ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just schoolId ->
+          teachers ^. TeacherSchoolId ==. val (Just schoolId)
+        Nothing ->
+          teachers ^. TeacherId ==. val teacherId
+    pure memberships
+```
+
+One abbreviation we seem to tolerate is `num` as in `numQuestionsAnswered`. There may be others.
+
+## JavaScript
+
+JavaScript uses `camelCase` for identifiers and `TitleCase` for classes and React components. For flow types, we generally append a `T` to the end of the type, e.g. `SkillT`. Names that are being ignored should be prefixed with an underscore (e.g. `_teacher`, `_standard`). It's a common mistake (especially for Haskellers) to use a lone `_` and accidentally overwrite underscore in that scope.
+
+## Problems
+
+Serialization of string-enums to JSON and postgres is all over the place. We have examples of `snake_case`, `camelCase`, `TitleCase`, and `ALLCAPS`.
+
+If we have DB-level enums that are not shared with the frontend, it makes sense to use the postgres convention of `snake_case`. However, if we're planning to pass enums to the frontend, then we should use whatever the JavaScript convention is. JavaScript is also using pretty much everything, so it's hard to say what we should be using. I (Parks) usually use `TitleCase` for JavaScript-only enums, but only because that mimics Haskell enums.
```

Comment:
If we have some enumerated column in a database entity, and spit it back to the frontend in the traditional way, `snake_case` will bleed into the frontend. This has a good chance of happening frequently, and without much thought.

I think it would be good to settle on a convention here. Couple of things to consider:

- When we pass an enum value back to the frontend, can we pre-process it into a different casing format? Can this be done by modifying typeclass instances?
- Can we do the same thing on the frontend by immediately converting `snake_case` values to `TitleCase`?
- Is it ever acceptable to use `camelCase` for enums?
- How much work would it take to:
  - Modify all database-level enums to be `TitleCase`?
  - Modify the API/Frontend bridge enums to be `snake_case`?

It would definitely be nice to settle on something here. I think at the very least we should aim to outlaw some naming schemes in specific contexts.

Reply:
The one problem here is using `JSON` encoding for database values that are not actually `JSON`. We should really stop doing that. If something is indead `JSONB` data then it should be stored in accorance with our `JSON` style guide. Otherwise we shouldn't leverage encoding functions that use `To/FromJSON`.

Reply:
I will be totally in converting as soon as receive information in the frontend to the format the frontend should use (camelCase). But it turn out hard to do because content we receive directly from JSON.
- Also the fancy api path route JS file is using snake case